src/test/java/org/springframework/integration/kafka/inbound/MessageDrivenAdapterTests.java

```diff
@@ -147,6 +157,106 @@ public void testInboundRecord() throws Exception {
 		adapter.stop();
 	}
 
+	@SuppressWarnings("deprecation")
+	@Test
+	public void testInboundRecordRetryRecover() throws Exception {
+		Map<String, Object> props = KafkaTestUtils.consumerProps("test4", "true", embeddedKafka);
+		props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+		DefaultKafkaConsumerFactory<Integer, String> cf = new DefaultKafkaConsumerFactory<Integer, String>(props);
+		ContainerProperties containerProps = new ContainerProperties(topic4);
+		KafkaMessageListenerContainer<Integer, String> container =
+				new KafkaMessageListenerContainer<>(cf, containerProps);
+		KafkaMessageDrivenChannelAdapter<Integer, String> adapter = new KafkaMessageDrivenChannelAdapter<>(container);
+		MessageChannel out = new DirectChannel() {
+
+			@Override
+			protected boolean doSend(Message<?> message, long timeout) {
+				throw new RuntimeException("intended");
+			}
+
+		};
+		adapter.setOutputChannel(out);
+		RetryTemplate retryTemplate = new RetryTemplate();
+		SimpleRetryPolicy retryPolicy = new SimpleRetryPolicy();
+		retryPolicy.setMaxAttempts(1);
+		retryTemplate.setRetryPolicy(retryPolicy);
+		QueueChannel errorChannel = new QueueChannel();
+		adapter.setRecoveryCallback(
+				new ErrorMessageSendingRecoverer(errorChannel, new RawRecordHeaderErrorMessageStrategy()));
```

Comment:
I guess there must be something like recommendation to use this `RawRecordHeaderErrorMessageStrategy` for the recovery. Since exactly that one is used out-of-the-box in the Adapter for raw error handling.